### DISCUSSION TITLE
Marshalling the `lpDesktop` field of the `STARTUPINFOW` structure

### DISCUSSION THREAD

**Discussion by vitkuz573** [microsoft/CsWin32#769] [Q&A] [answered]
Is it possible to marshall the `lpDesktop` field of the `STARTUPINFOW` structure to string?

Right now the type of this field is `PWSTR` and this makes it very difficult to use

`var si = new STARTUPINFOW();`
`si.lpDesktop = @"winsta0\" + desktopName; // Failed to implicitly convert string type to Windows.Win32.Foundation.PWSTR`

**Comment by KalleOlaviNiemitalo** [comment]
Is there a risk that a future version of Windows could overload lpDesktop with some new flag in dwFlags, similar to how hStdInput can be either a handle or a hot key?  That would then conflict with any automatic string marshalling.

Although there is STARTUPINFOEX nowadays, the GetStartupInfo function doesn't appear to support it, so any new settings that need to be readable by application code (as opposed to operating system code) in the new process might still be placed in STARTUPINFO.

**Comment by mikebattista** [comment]
@aarnott is this a C#/Win32 question?

**Comment by AArnott** [comment, marked as answer]
There are two sides to your question:

> Is it possible to marshall the lpDesktop field of the STARTUPINFOW structure to string?

Yes. This is easy: just call `PWSTR.ToString()`.

But your code snippet asks a different question:

```cs
var si = new STARTUPINFOW();
si.lpDesktop = @"winsta0\" + desktopName;
```

This requires marshaling a string to a PWSTR, which is a bit trickier, because now you have to manage pinned memory while you use the struct. Here's how to do it:

```cs
fixed (char* pDesktopName = @"winsta0\" + desktopName)
{
    STARTUPINFOW si = default;
    si.lpDesktop = pDesktopName;

    // Also *pass* STARTUPINFOW to any function that needs it within this fixed block so the pointer remains valid.
}
```

As this uses pointers, you'll need to be in an `unsafe` block or a member with an `unsafe` modifier.
But don't let that scare you. `unsafe` is much easier than a whole bunch of IntPtr and Marshal class arithmetic to avoid pointers.

**Reply by vitkuz573** [reply]
Thank you very much for the explanations!
